### ISSUE TITLE
does not index into shape Error

### ISSUE THREAD

**Issue by afcruzs** [emadboctorx/yolo-tf2#3]
While trying to train with my own dataset I got the following error:

```
  File "../Helpers/utils.py", line 57, in wrapper
    result = func(*args, **kwargs)
  File "/home/jupyter/mask-faces/train/od/Main/trainer.py", line 544, in train
    validation_data=valid_dataset,
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 66, in _method_wrapper
    return method(self, *args, **kwargs)
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/keras/engine/training.py", line 848, in fit
    tmp_logs = train_function(iterator)
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 580, in __call__
    result = self._call(*args, **kwds)
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/def_function.py", line 611, in _call
    return self._stateless_fn(*args, **kwds)  # pylint: disable=not-callable
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 2420, in __call__
    return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 1665, in _filtered_call
    self.captured_inputs)
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 1746, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager))
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/function.py", line 598, in call
    ctx=ctx)
  File "/opt/conda/lib/python3.7/site-packages/tensorflow/python/eager/execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.InvalidArgumentError:   [_Derived_]  indices[2] = [2, 10, 6, 2] does not index into shape [8,10,10,3,6]
         [[{{node TensorScatterUpdate}}]]
         [[PartitionedCall_1]]
         [[IteratorGetNext]] [Op:__inference_train_function_29922]

Function call stack:
train_function -> train_function -> train_function -> train_function
```

My training code looks as follows:

```
    tr = Trainer(
            input_shape=(160, 160, 3),
            classes_file=args.classes_file,
            image_width=640,  # The original image width
            image_height=480,   # The original image height
            image_folder=args.dataset_folder,
            train_tf_record=None,
            valid_tf_record=None
    )

    dataset_conf = {
                'relative_labels': args.relative_labels,
                'dataset_name': DATASET_NAME,
                'test_size': 0.1,
                'sequences': preset_1,  # check Config > augmentation_options.py
                'augmentation': args.augmentation,
    }

    anchors_conf = {
                'anchor_no': 9,
                'relative_labels':  args.relative_labels
    }

    tr.train(epochs=100, 
            batch_size=8, 
            learning_rate=1e-3, 
            dataset_name=DATASET_NAME, 
            merge_evaluation=False,
            min_overlaps=0.5,
            new_dataset_conf=None,
            new_anchors_conf=None,
            weights=None)
```

I extended the code a bit to receive the image_folder on the Trainer ctor (fork [here](https://github.com/afcruzs/yolov3-keras-tf2)), but I think that shouldn't change anything. 

Any idea what could this be?

**Comment by emadboctorx** [comment]
Sometimes this crazy shapes problem occurs because of the augmentation, photos flipping not in the [90, 180] angles however it shouldn't happen with `preset_1` if you're using it, anyway try to reproduce without image augmentation and see if the problem is still there, then try reproducing the problem with my version and if all fails, state all the steps to reproduce the error and indicate the number of classes, weights you're using and a link to your dataset if possible.

**Comment by MikeMpapa** [comment]
Thanks for this great project! 

I do get the same error as @afcruzs when trying to train on a custom dataset without augmentation and using your implementation (no code adjustments). My task has 5 classes and I used random weights. Here is my setup:

  trainer = Trainer(
        input_shape=(416, 416, 3),
        classes_file='classes.txt',
        image_width=640,  # The original image width
        image_height=512  # The original image height
    )

    dataset_conf = {
        'relative_labels': 'YOLO_input.csv',
        'dataset_name': 'MyData',
        'test_size': 0.2,
        #'sequences': preset_1,  # check Config > augmentation_options.py
        'augmentation': False,
    }

    anchors_conf = {
                    'anchors_no': 9,
                    'relative_labels':  'YOLO_input.csv'
    }

    trainer.train(epochs=100,
             batch_size=8,
             learning_rate=1e-3,
             dataset_name='MyData',
             merge_evaluation=False,
             min_overlaps=0.5,
             new_dataset_conf=dataset_conf,  # check step 5
             new_anchors_conf=anchors_conf,  # check step 6
             #  weights='/path/to/weights'  # If you're using DarkNet weights or resuming training
             )


When trying augmentation the process also breaks while augmenting with the following error:
  
File "/home/mpapakos/.conda/envs/YOLO_Thermal/lib/python3.7/site-packages/pandas/core/groupby/groupby.py", line 686, in get_group
    raise KeyError(name)
KeyError: 'aug-8188395-sequence-7-06147.png'

NOTE1: You might want to add imagecorruptions in your dependencies as it is not included in the standard imgaug package

NOTE2: There is a misspell in Readme that beaks the code:
anchors_conf = {
                    'anchors_no': 9,
                    'relative_labels':  'YOLO_input.csv'
    }
trainer.py line 141, looks for 'anchor_no', not 'anchors_no'

**Comment by emadboctorx** [comment]
@MikeMpapa the augmentation key error means you have some image in the csv file that is not an actual path check the photos, you will not find this image name. Thanks for letting me know about the Readme, i'll fix it and regarding the shapes problem, most probably it's a dataset problem, make sure the annotations are correct, you can use my other repo labelpix for examining the dataset and check for bad annotations

**Comment by MikeMpapa** [comment]
Thanks! I will check out labelpix for annotation problems.

 Regarding the augmentation issue not sure I understood. Image '06147.png' (which is the original filename) exists both in my csv file and as a path (I can see the pic in the directory). I can also see  'aug-8188395-sequence-7-06147.png' being generated in the same directory, which is the filename reported in the error. 

One last question, does your implementation support overlapping bboxes? 
Thanks again!

**Comment by emadboctorx** [comment]
Yes it does support overlapping boxes and i recommend to delete photos and retry the training and see if it occurs again, open another issue and i'll look into fixing whatever that might be causing the problem

**Comment by azeembootwala** [comment]
I am facing the same issue though I do not use any augmentations

`2020-06-08 13:03:15.601089: I tensorflow/core/profiler/lib/profiler_session.cc:159] Profiler session started.
2020-06-08 13:03:15.605503: I tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1363] Profiler found 1 GPUs
2020-06-08 13:03:15.609718: W tensorflow/stream_executor/platform/default/dso_loader.cc:55] Could not load dynamic library 'cupti64_101.dll'; dlerror: cupti64_101.dll not found
2020-06-08 13:03:15.622711: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1408] function cupti_interface_->Subscribe( &subscriber_, (CUpti_CallbackFunc)ApiCallback, this)failed with error CUPTI could not be loaded or symbol could not be found.
2020-06-08 13:03:15.638168: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1447] function cupti_interface_->ActivityRegisterCallbacks( AllocCuptiActivityBuffer, FreeCuptiActivityBuffer)failed with error CUPTI could not be loaded or symbol could not be found.
2020-06-08 13:03:15.650571: E tensorflow/core/profiler/internal/gpu/cupti_tracer.cc:1430] function cupti_interface_->EnableCallback( 0 , subscriber_, CUPTI_CB_DOMAIN_DRIVER_API, cbid)failed with error CUPTI could not be loaded or symbol could not be found.
WARNING:tensorflow:Model failed to serialize as JSON. Ignoring... can't pickle _thread.RLock objects
Epoch 1/3
2020-06-08 13:03:35.997088: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at scatter_nd_op.cc:217 : Invalid argument: indices[0] = [0, 20, 16, 2] does not index into shape [2,13,13,3,6]
Traceback (most recent call last):
  File "run_ab.py", line 56, in <module>
    train()
  File "run_ab.py", line 51, in train
    new_anchors_conf= anchors_conf
  File "E:\Projects\YOLOV3\yolov3-keras-tf2\Helpers\utils.py", line 63, in wrapper
    result = func(*args, **kwargs)
  File "E:\Projects\YOLOV3\yolov3-keras-tf2\Main\trainer.py", line 522, in train
    validation_data=valid_dataset,
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\keras\engine\training.py", line 66, in _method_wrapper
    return method(self, *args, **kwargs)
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\keras\engine\training.py", line 848, in fit
    tmp_logs = train_function(iterator)
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\def_function.py", line 580, in __call__
    result = self._call(*args, **kwds)
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\def_function.py", line 644, in _call
    return self._stateless_fn(*args, **kwds)
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\function.py", line 2420, in __call__
    return graph_function._filtered_call(args, kwargs)  # pylint: disable=protected-access
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\function.py", line 1665, in _filtered_call
    self.captured_inputs)
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\function.py", line 1746, in _call_flat
    ctx, args, cancellation_manager=cancellation_manager))
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\function.py", line 598, in call
    ctx=ctx)
  File "D:\Anaconda3\envs\tf_2.0\lib\site-packages\tensorflow\python\eager\execute.py", line 60, in quick_execute
    inputs, attrs, num_outputs)
tensorflow.python.framework.errors_impl.InvalidArgumentError:   [_Derived_]  indices[0] = [0, 20, 16, 2] does not index into shape [2,13,13,3,6]
         [[{{node TensorScatterUpdate}}]]
         [[PartitionedCall]]
         [[IteratorGetNext]] [Op:__inference_train_function_29657]

Function call stack:
train_function -> train_function -> train_function -> train_function

2020-06-08 13:03:36.041192: W tensorflow/core/framework/op_kernel.cc:1753] OP_REQUIRES failed at scatter_nd_op.cc:217 : Invalid argument: indices[0] = [0, 18, 6, 2] does not index into shape [2,13,13,3,6]`

My training code does not use augmentations and is as follows:
`t = Trainer(
        input_shape = (416,416,3),
        model_configuration = './Config/yolo3.cfg',
        classes_file = 'classes.txt',
        image_width = 1000,
        image_height = 10000
    )


    dataset_conf = {
        'dataset_name': 'dataset_name',
        'relative_labels': args.label_file,
        'test_size': 0.1,
    }

    anchors_conf = {
        'anchor_no': 9,
        'relative_labels': args.label_file
    }


    t.train(
        epochs = 3,
        batch_size = 2,
        learning_rate=1e-2,
        dataset_name='dataset_name',
        shuffle_buffer = 2,
        evaluation_workers = 1,
        merge_evaluation=False,
        new_dataset_conf= dataset_conf,
        new_anchors_conf= anchors_conf
    )`

**Comment by emadboctorx** [comment]
@azeembootwala There are a bunch of possible reasons that could lead to these shape mismatch problems, check your dataset for possible bad annotations, check the shapes in the error message and find which tensors have these shapes indicated, minimize the dataset to like 10 examples and see if the problem still occurs and use the default settings, do not use your preferences while debugging, there is no one solution fits all kind of fix, the code runs perfectly fine on several datasets that I tested myself, these shape mismatches indicate that there is one possible of many factors that is problematic, which my best guess is dataset specific but can be any other factor

**Comment by MikeMpapa** [comment]
@azeembootwala for me the error was fixed after correcting the coordinates. Make sure you either use the csv with the relative coordinates or the xml VOC with the real coordinates. Check here: https://github.com/emadboctorx/yolov3-keras-tf2/issues/6. Labelpix also helped a lot to double-check the bboxes before starting the training.

**Comment by emadboctorx** [comment]
@MikeMpapa do you have a suggestion about modifications that can be introduced to the current code to maybe make some pre-check to avoid this particular error? I myself think it's unavoidable because most of the error messages you get in machine learning are about incompatible shapes or what do you think?

**Comment by emadboctorx** [comment]
I closed this issue as it has stalled, feel free to re-open it when you need to
